# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [1]:

import warnings
import numpy as np
import sqlite3
import pandas as pd
warnings.simplefilter('ignore')

C:\Users\Anthony Delgado\AppData\Local\Temp\ipykernel_131940\476482988.py:4: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [2]:
global con
def dataframe() : 
    try : 
        global con 
        con = sqlite3.connect("Assets/Week 2/go_sales.sqlite")
        print("Connected to sqlite.")
    except sqlite3.Error as error: 
        print("Failed to read data from sqlite table", error)
    

dataframe()
print(con)

Connected to sqlite.


Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [3]:
global product
product = pd.read_sql_query("SELECT * FROM {}".format("product"),con)

global product_type
product_type = pd.read_sql_query("SELECT * FROM {}".format("product_type"),con)

global product_line
product_line = pd.read_sql_query("SELECT * FROM {}".format("product_line"),con)

global sales_staff
sales_staff = pd.read_sql_query("SELECT * FROM {}".format("sales_staff"),con)

global sales_branch
sales_branch = pd.read_sql_query("SELECT * FROM {}".format("sales_branch"),con)

global retailer_site
retailer_site = pd.read_sql_query("SELECT * FROM {}".format("retailer_site"),con)

global country
country = pd.read_sql_query("SELECT * FROM {}".format("country"),con)

global order_header
order_header = pd.read_sql_query("SELECT * FROM {}".format("order_header"),con)

global order_details
order_details = pd.read_sql_query("SELECT * FROM {}".format("order_details"),con)

global target
target = pd.read_sql_query("SELECT * FROM {}".format("sales_targetdata"),con)

global returned_item
returned_item = pd.read_sql_query("SELECT * FROM {}".format("returned_item"),con)

global return_reason
return_reason = pd.read_sql_query("SELECT * FROM {}".format("return_reason"),con)



Columntypes converteren

In [4]:
product["PRODUCTION_COST"] = product["PRODUCTION_COST"].astype(float)
product["MARGIN"] = product["MARGIN"].astype(float)
order_details["UNIT_SALE_PRICE"] = order_details["UNIT_SALE_PRICE"].astype(float)
order_details["UNIT_PRICE"] = order_details["UNIT_PRICE"].astype(float)
order_details["QUANTITY"] = order_details["QUANTITY"].astype(int)
returned_item["RETURN_QUANTITY"] = returned_item["RETURN_QUANTITY"].astype(int)
# target["SALES_YEAR"] = target["SALES_YEAR"].astype(int)

# print(product.dtypes)
# print()
# print(order_details.dtypes)
# print()
# print(returned_item.dtypes)
# print()
print(target.dtypes)

Id                   int64
SALES_STAFF_CODE    object
SALES_YEAR          object
SALES_PERIOD        object
RETAILER_NAME       object
PRODUCT_NUMBER      object
SALES_TARGET        object
RETAILER_CODE       object
TRIAL888            object
dtype: object


Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [5]:
def production_cost_Dataframe() :
    product["PRODUCTION_COST"] = product["PRODUCTION_COST"].astype(float)
    productionconstresult_filter = (product["PRODUCTION_COST"] > 50) & (product["PRODUCTION_COST"] < 100)
    productioncostresult = product.loc[(productionconstresult_filter), ['PRODUCT_NAME','PRODUCTION_COST']]
    print(productioncostresult.count())
    display(productioncostresult)
    # # print(product.keys())
production_cost_Dataframe()

PRODUCT_NAME       23
PRODUCTION_COST    23
dtype: int64


PRODUCT_NAME  PRODUCTION_COST
13                Blue Steel Putter            60.00
14            Blue Steel Max Putter            81.80
17              Course Pro Golf Bag            80.00
24                  Hibernator Lite            60.00
25                       Hibernator            86.00
28  Hibernator Self - Inflating Mat            78.55
31              Hibernator Camp Cot            65.33
32     Canyon Mule Climber Backpack            62.50
56          Granite Climbing Helmet            52.86
58            Husky Harness Extreme            55.00
69                   Granite Hammer            57.13
71        TrailChef Deluxe Cook Set            85.11
81                        Polar Ice            73.33
83                     Polar Sports            80.00
84                       Polar Wave            60.00
85                    Polar Extreme            72.50
88                     Edge Extreme            80.00
91                        Seeker 35            79.19
92                        Seeker 50            92.58
93           TrailChef Double Flame            75.00
94                   Seeker Extreme            94.12
97                   Glacier Deluxe            56.25
98                      Glacier GPS            78.55

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [6]:
def product_margin_between() :    
    product_margin_filter = (product["MARGIN"]< 0.20) | (product['MARGIN']>0.60)
    product_margin_result = product.loc[(product_margin_filter),['PRODUCT_NAME','MARGIN']]
    print(product_margin_result.count())
    display(product_margin_result)
product_margin_between()

PRODUCT_NAME    7
MARGIN          7
dtype: int64


PRODUCT_NAME  MARGIN
29                 Hibernator Pad    0.17
32   Canyon Mule Climber Backpack    0.17
100             BugShield Natural    0.67
101               BugShield Spray    0.67
102         BugShield Lotion Lite    0.70
103              BugShield Lotion    0.63
105             BugShield Extreme    0.63

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [7]:
def country_francs() :
    country_currency_filter = (country["CURRENCY_NAME"] == "francs")
    country_currency = country.loc[(country_currency_filter),["COUNTRY"]]
    country_currency_sorted = country_currency.sort_values("COUNTRY")
    print(country_currency_sorted.count())
    display(country_currency_sorted)
country_francs()


COUNTRY    3
dtype: int64


COUNTRY
7       Belgium
0        France
19  Switzerland

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [8]:
def introduction_margins() :
    introduction_margin_filter = product["MARGIN"]>0.50
    introduction_margin = product.loc[(introduction_margin_filter),["INTRODUCTION_DATE"]].drop_duplicates("INTRODUCTION_DATE")
    print(introduction_margin.count())
    display(introduction_margin)
introduction_margins()

INTRODUCTION_DATE    7
dtype: int64


INTRODUCTION_DATE
4          10-12-2019
6          18-12-2019
7          27-12-2019
9           13-1-2020
14         15-12-2019
79         26-10-2016
100         15-2-2011

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

In [9]:
def first_addres() :
    sales_branch_filter = (sales_branch["ADDRESS2"].notna()) & (sales_branch["REGION"].notna())
    sales_branch_location = sales_branch.loc[(sales_branch_filter),["ADDRESS1","CITY"]]
    print(sales_branch_location.count())
    display(sales_branch_location)
first_addres()

ADDRESS1    7
CITY        7
dtype: int64


ADDRESS1              CITY
10  Prol. Paseo de la Reforma No. 51  Distrito Federal
11               202-2-3 Hyakunincho             Tokyo
12                     543-225 Asahi        Osaka City
13                  2315 Queen's Ave         Melbourne
15             Avenida Paulista, 333         São Paulo
21                    3 Albany Court        Birmingham
27               Singelgravenplein 4         Amsterdam

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [10]:
def country_dollars() :
    country_dollar_filter = (country["CURRENCY_NAME"].str.contains("dollar"))
    # print(country_dollar_filter)
    country_dollar_currency = country.loc[(country_dollar_filter),["COUNTRY"]]
    print(country_dollar_currency.count())
    display(country_dollar_currency)
country_dollars()

COUNTRY    4
dtype: int64


COUNTRY
3          Taiwan
6       Australia
14  United States
15         Canada

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [11]:
country_postal_filter = ( sales_branch["POSTAL_ZONE"].str[0]=="D")
country_postal = sales_branch.loc[(country_postal_filter),["ADDRESS1","ADDRESS2","CITY"]]
country_postal_sorted = country_postal.sort_values("ADDRESS2")
print(country_postal_sorted.count())
display(country_postal_sorted)

ADDRESS1    2
ADDRESS2    0
CITY        2
dtype: int64


ADDRESS1 ADDRESS2     CITY
0    Schwabentor 35     None  Hamburg
1  Leopoldstraße 36     None  München

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [12]:
print(len(returned_item))

706


Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [13]:
print(len(pd.unique(sales_branch["REGION"])))

16


Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [14]:
# print(product.keys())
min = product["MARGIN"].min()
avg = product["MARGIN"].mean()
max = product["MARGIN"].max()


df = pd.DataFrame(
     {
        "Min":min,
        "Avg":avg,
        "Max":max
     }, index=[0]
    )
display(df)


Min       Avg  Max
0  0.17  0.396609  0.7

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [15]:
print(len(sales_branch[sales_branch["ADDRESS2"].isna()]))

19


Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [16]:
def calculate_average_sales_price() : 
    korting_filter = (order_details["UNIT_SALE_PRICE"] < order_details["UNIT_PRICE"])
    # print(korting_filter)
    gemiddelde_prijs_met_korting = order_details.loc[(korting_filter),:]
    avg_prijs = gemiddelde_prijs_met_korting["UNIT_SALE_PRICE"].mean()
    print(avg_prijs)


calculate_average_sales_price()

177.99788862151942


Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [17]:
def medewerkers_per_functie() : 
    medewerkers_filter = pd.unique(sales_staff["POSITION_EN"])
    # print(medewerkers_filter)
    # display(sales_staff["POSITION_EN"].value_counts())
    medewerker_soorten = pd.DataFrame(sales_staff.groupby("POSITION_EN")["SALES_STAFF_CODE"].count())
    display(medewerker_soorten)

medewerkers_per_functie()

SALES_STAFF_CODE
POSITION_EN                                   
Branch Manager                              19
District Manager                             4
General Manager                              1
Level 1 Sales Representative                13
Level 2 Sales Representative                39
Level 3 Sales Representative                24
Regional Manager                             2

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [18]:
def medewerkers_per_telefoonnummer_4() : 
    count = pd.DataFrame(sales_staff["WORK_PHONE"].value_counts())
    count_filter = (count["count"]>4)
    count_filtered = count.loc[(count_filter),:]
    print(count_filtered.count())
    display(count_filtered)
    

    # print(len(sales_staff))
    # medewerkers_filtered = sales_staff.loc[medewerker_filter,["WORK_PHONE"]]
    # display(medewerkers_filtered)
    # print(sales_staff["WORK_PHONE"].value_counts())

medewerkers_per_telefoonnummer_4()

count    10
dtype: int64


count
WORK_PHONE                
1 (305) 557-4810         6
1 (617) 268-6754         6
+31 (0)20 692 93 94      5
1 (403) 232-5986         5
+(61) 03 2982  4242      5
+(44) 121 3505267        5
1 (310) 281-5722         5
+33 1 68 94 52 20        5
1 (206) 292-0012         5
1 (416) 493-5595         5

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [19]:
def location_details_dutchCustomers() : 
    client_country = pd.merge(retailer_site, country, left_on="COUNTRY_CODE", how="left",right_on="COUNTRY_CODE")
    country_filter = (client_country["COUNTRY"] == "Netherlands")
    client_country_filtered = client_country.loc[(country_filter), ["ADDRESS1","CITY"]] 

    print(client_country_filtered.count())
    display(client_country_filtered)

location_details_dutchCustomers()

ADDRESS1    20
CITY        20
dtype: int64


ADDRESS1          CITY
164                   Beets 36         Beets
165               Perestraat 1      Deventer
166  Laan van Meerdervoort 966      Den Haag
167             Kalfjeslaan 47    Amstelveen
168               Startbaan 11    Amstelveen
169      Noorderspoorsingel 28     Groningen
170              Hogehilweg 13     Amsterdam
171             Spoorstraat 43    Varsseveld
173             Westhavenweg 8     Amsterdam
174              Westzeedijk 4        Muiden
175            Gooiergracht 91         Laren
176            Hoofdstraat 150   Pieterburen
177                Olde Diek 2         Laren
178     Amsterdamsestraatweg 8      Maarssen
179        Krayenhoffstraat 11     Amsterdam
180         Overschiestraat 19     Amsterdam
181              Havenbuurt 12        Marken
182               Duivendaal 1    Wageningen
184              Voorstraat 37  Nieuweschans
185         Stadhouderslaan 18     Rotterdam

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

In [20]:
def eyewear_products_insight() : 
    productType_products = pd.merge(product, product_type, left_on="PRODUCT_TYPE_CODE", how="left",right_on="PRODUCT_TYPE_CODE")
    eyewear_filter = (productType_products["PRODUCT_TYPE_EN"] == "Eyewear")
    eyewear_products = productType_products.loc[(eyewear_filter),["PRODUCT_NAME"]]
    
    print(eyewear_products.count())
    display(eyewear_products)

eyewear_products_insight()

PRODUCT_NAME    5
dtype: int64


PRODUCT_NAME
80      Polar Sun
81      Polar Ice
83   Polar Sports
84     Polar Wave
85  Polar Extreme

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

In [21]:
def branch_sold_to_unique_client_location() : 
    clientLocation_sales = pd.merge(retailer_site, order_header, left_on="RETAILER_SITE_CODE", how="left",right_on="RETAILER_SITE_CODE")
    clientLocation_sales_branchManager = pd.merge(clientLocation_sales,sales_staff,left_on="SALES_STAFF_CODE",how="left",right_on="SALES_STAFF_CODE")
    BranchMang_filter = (clientLocation_sales_branchManager["POSITION_EN"] == "Branch Manager")
    clientLocation_sales_branchManager_FILTERED = clientLocation_sales_branchManager.loc[(BranchMang_filter),["ADDRESS1","FIRST_NAME","LAST_NAME","POSITION_EN"]]
    # print(clientLocation_sales_branchManager_UNIQUE.count())
    display(clientLocation_sales_branchManager_FILTERED.loc[:,:].drop_duplicates("ADDRESS1"))

branch_sold_to_unique_client_location()

ADDRESS1 FIRST_NAME LAST_NAME     POSITION_EN
3058  Alameda Santos, 9876     Bayard     Lopes  Branch Manager

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

In [22]:
def list_salesmen_if_sold() : 
    sales_salesStaff = pd.merge(order_header, sales_staff, left_on="SALES_STAFF_CODE",how="outer",right_on="SALES_STAFF_CODE")
    position_salesDate_filter = ( sales_salesStaff["POSITION_EN"].str.contains("Manager"))
    #zelfs met outer join zijn er veel lege waardes. Deze uitwerking moet vergeleken worden met dat van anderen. 
    position_salesDate = sales_salesStaff.loc[(position_salesDate_filter),["POSITION_EN","ORDER_DATE"]].dropna()
    display(position_salesDate)

list_salesmen_if_sold()

POSITION_EN  ORDER_DATE
5325  Branch Manager  2022-09-09
5326  Branch Manager  2022-10-31
5327  Branch Manager  2022-12-13
5328  Branch Manager  2022-03-13
5329  Branch Manager  2022-01-09

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

In [23]:
def listProducts_quantitySoldGreaterThan750() : 
    product_productTypes = pd.merge(product,product_type,left_on="PRODUCT_TYPE_CODE",how="left",right_on="PRODUCT_TYPE_CODE")
    productTypeSales = pd.merge(product_productTypes,order_details,left_on="PRODUCT_NUMBER",how="left",right_on="PRODUCT_NUMBER")
    quantity_filter = (productTypeSales["QUANTITY"]>750)
    quantity_sold = productTypeSales.loc[(quantity_filter),["PRODUCT_NAME","PRODUCT_TYPE_EN"]].drop_duplicates()

    print(quantity_sold.count())
    display(quantity_sold)

listProducts_quantitySoldGreaterThan750()

PRODUCT_NAME       9
PRODUCT_TYPE_EN    9
dtype: int64


PRODUCT_NAME    PRODUCT_TYPE_EN
8616                Star Peg              Tents
15975              Firefly 2           Lanterns
37903        BugShield Spray  Insect Repellents
38313  BugShield Lotion Lite  Insect Repellents
38770       BugShield Lotion  Insect Repellents
39618      BugShield Extreme  Insect Repellents
40754         Sun Shelter 15          Sunscreen
41106         Sun Shelter 30          Sunscreen
41681             Sun Shield          Sunscreen

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

In [24]:
def listProducts_DiscountGreaterThan40() : 
    products_ordered = pd.merge(order_details,product,left_on="PRODUCT_NUMBER",how="left",right_on="PRODUCT_NUMBER")
    productNames_filter = ((products_ordered["UNIT_PRICE"] - products_ordered["UNIT_SALE_PRICE"])/products_ordered["UNIT_PRICE"])>0.40
    productNames = products_ordered.loc[(productNames_filter),["PRODUCT_NAME"]].drop_duplicates()
    
    print(productNames.count())
    display(productNames)

listProducts_DiscountGreaterThan40()

PRODUCT_NAME    8
dtype: int64


PRODUCT_NAME
5594          BugShield Lotion Lite
6826              BugShield Extreme
9413              BugShield Natural
12276                Sun Shelter 30
14200              BugShield Lotion
14985               BugShield Spray
15124  Hailstorm Titanium Woods Set
20108                Sun Shelter 15

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

In [25]:
def returnReason_ItemReturnedGreaterThan90Percent() :
    product_order = pd.merge(product, order_details, left_on="PRODUCT_NUMBER",how="left",right_on="PRODUCT_NUMBER")
    order_return = pd.merge(product_order, returned_item, left_on="ORDER_DETAIL_CODE",how="left",right_on="ORDER_DETAIL_CODE")
    order_return_reason = pd.merge(order_return, return_reason, left_on="RETURN_REASON_CODE",how="left",right_on="RETURN_REASON_CODE")
    returnQuantity_filter = (order_return_reason["RETURN_QUANTITY"]/order_return_reason["QUANTITY"])>0.9



    returnReason = order_return_reason.loc[(returnQuantity_filter),["RETURN_DESCRIPTION_EN"]].drop_duplicates("RETURN_DESCRIPTION_EN")

    
    display(returnReason)

returnReason_ItemReturnedGreaterThan90Percent()

RETURN_DESCRIPTION_EN
168   Wrong product ordered
313  Unsatisfactory product
618   Wrong product shipped

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [26]:
def products_per_type() : 
    product_productTypes = pd.merge(product,product_type,left_on="PRODUCT_TYPE_CODE",how="left",right_on="PRODUCT_TYPE_CODE")
    products_per_type = pd.DataFrame(product_productTypes.groupby("PRODUCT_TYPE_EN")["PRODUCT_NUMBER"].count())
    print(products_per_type.count())
    display(products_per_type)

products_per_type()

PRODUCT_NUMBER    21
dtype: int64


PRODUCT_NUMBER
PRODUCT_TYPE_EN                     
Binoculars                         4
Climbing Accessories               7
Cooking Gear                      10
Eyewear                            5
First Aid                          5
Golf Accessories                   4
Insect Repellents                  5
Irons                              4
Knives                             5
Lanterns                          12
Navigation                         4
Packs                              6
Putters                            3
Rope                               4
Safety                             4
Sleeping Bags                      7
Sunscreen                          5
Tents                              6
Tools                              6
Watches                            5
Woods                              4

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

In [27]:
def client_based_in_country() : 
    country_clients = pd.merge(retailer_site,country,left_on="COUNTRY_CODE",how="left",right_on="COUNTRY_CODE")
    clients_per_country = pd.DataFrame(country_clients.groupby("COUNTRY")["RETAILER_CODE"].count())
    print(clients_per_country.count())
    display(clients_per_country)
    


client_based_in_country()

RETAILER_CODE    21
dtype: int64


RETAILER_CODE
COUNTRY                      
Australia                  10
Austria                    10
Belgium                    10
Brazil                      1
Canada                     31
China                      10
Denmark                     6
Finland                     5
France                     29
Germany                    32
Italy                      12
Japan                      25
Korea                      10
Mexico                     10
Netherlands                20
Spain                       9
Sweden                     12
Switzerland                13
Taiwan                     10
United Kingdom             25
United States             101

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

In [28]:
def quantitySold_ofTypeCookingGear() : 
    product_productTypes = pd.merge(product,product_type,left_on="PRODUCT_TYPE_CODE",how="left",right_on="PRODUCT_TYPE_CODE")
    sales_product_productTypes = pd.merge(order_details, product_productTypes, left_on="PRODUCT_NUMBER",how="left",right_on="PRODUCT_NUMBER")
    type_filter = sales_product_productTypes["PRODUCT_TYPE_EN"] == "Cooking Gear"
    sales_product_productTypes_filtered = sales_product_productTypes.loc[(type_filter),:]
    columns = ["PRODUCT_TYPE_EN","PRODUCT_NAME"]
    count = sales_product_productTypes_filtered.groupby(columns).aggregate({"QUANTITY":"sum","UNIT_PRICE":"mean"})

    display(count)

quantitySold_ofTypeCookingGear()

QUANTITY  UNIT_PRICE
PRODUCT_TYPE_EN PRODUCT_NAME                                   
Cooking Gear    TrailChef Canteen             31712   12.178329
                TrailChef Cook Set            12700   52.605173
                TrailChef Cup                 31834    7.137756
                TrailChef Deluxe Cook Set      3744  123.537733
                TrailChef Double Flame        14510  136.601562
                TrailChef Kettle               9900   12.375391
                TrailChef Kitchen Kit         25964   22.744308
                TrailChef Single Flame        10614   65.090826
                TrailChef Utensils            15928   17.731438
                TrailChef Water Bag           41770    6.181641

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2006. (4 kolommen, 4 rijen) 

~~*deze opdracht kan niet gemaakt worden. De tabel SalesTARGETData heeft alleen jaren 2020,2021 en 2022 in de kolom SALES_YEAR*~~
*Dit moet kennelijk het hoogste verkooptarget uit 2022 zijn.*

In [29]:
def salesStaff_Amsterdam_2006Targets() :
    staff_branch = pd.merge(sales_staff,sales_branch,left_on="SALES_BRANCH_CODE",how="left",right_on="SALES_BRANCH_CODE")
    staff_branch_targets = pd.merge(staff_branch,target,left_on="SALES_STAFF_CODE",how="left",right_on="SALES_STAFF_CODE")
    city_filter = (staff_branch_targets["CITY"] == "Amsterdam")
    year_filter = (staff_branch_targets["SALES_YEAR"] == "2022")
    staff_branch_targets_cityFiltered = staff_branch_targets.loc[(city_filter),["FIRST_NAME","LAST_NAME","SALES_YEAR","SALES_TARGET"]]
    staff_branch_targets_yearFiltered = staff_branch_targets_cityFiltered.loc[(year_filter),:]
    staff_branch_targets_aggregate = staff_branch_targets_yearFiltered.groupby(["FIRST_NAME","LAST_NAME"]).aggregate(Max_Target=("SALES_TARGET","max"),MIN_TARGET=("SALES_TARGET","min"))
    


    print(staff_branch_targets_aggregate.count())
    display(staff_branch_targets_aggregate)

salesStaff_Amsterdam_2006Targets()

Max_Target    4
MIN_TARGET    4
dtype: int64


Max_Target MIN_TARGET
FIRST_NAME LAST_NAME                             
Belinda    Jansen-Velasquez        963       1005
Ellen      Shapiro                 986       1004
Jan        Haverkamp               993       1048
Maria      Laponder               9956       1000

*Ik begrijp niet goed wat de bedoeling is met deze opdracht. Er zijn meerdere vestigingen per stad, van meerdere verkopers.*

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

In [30]:
def retailerSites_perCountry() : 
    branch_country = pd.merge(sales_branch, country, left_on="COUNTRY_CODE",how="left",right_on="COUNTRY_CODE")
    branch_country_retailer = pd.merge(branch_country, retailer_site, left_on="COUNTRY_CODE",how="left",right_on="COUNTRY_CODE")
    grouped_result = branch_country_retailer.groupby(["COUNTRY","CITY_y"]).aggregate(klanten=("CITY_y","count"))


    display(grouped_result)

retailerSites_perCountry()

klanten
COUNTRY       CITY_y              
Australia     Brisbane           1
              Cairns             2
              Melbourne          2
              NewCastle          2
              Perth              1
...                            ...
United States Seattle            8
              Springfield        4
              Topeka             4
              Tucson             4
              Washington         4

[229 rows x 1 columns]

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

In [31]:
def salesStaff_WithoutSales() : 
    salesperson = order_header["SALES_STAFF_CODE"]
    noSales_saleperson = sales_staff.loc[(~sales_staff["SALES_STAFF_CODE"].isin(salesperson)),["FIRST_NAME","LAST_NAME"]]
    print(noSales_saleperson.count())
    display(noSales_saleperson)

salesStaff_WithoutSales()

FIRST_NAME    25
LAST_NAME     25
dtype: int64


FIRST_NAME      LAST_NAME
7      Walter         Taylor
8    Penelope        Tamrine
11      Polly         Linton
19      Gracy        Gellens
26       Else         Mörike
27      Frank          Fuchs
30      Fritz         Hirsch
31       Jörg          Kunze
33      Maria      Iacobucci
38       Kick        Kalkman
43      Karin      Bergström
44      Denis           Pagé
47      Sally          White
49      Frank        Bretton
50    Georges  Saint-Germain
56       Alex      Rodriguez
60   Samantha          Floyd
65      Julie          Olsen
69      Corey         Wright
76    Estelle        Johnson
77       Greg        Belding
80      Julie        Béchard
84      Lucía          Reyna
89     Kazumi        Uragome
91      Alice         Walter

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

In [32]:
def productMarginLowerThanAverage() : 
    lowermargin = product.loc[(product["MARGIN"]<product["MARGIN"].mean()),:]
    lowermargincount = lowermargin["PRODUCT_NUMBER"].count()
    lowermarginmean = lowermargin["MARGIN"].mean()
    result = pd.Series( [ lowermargincount,lowermarginmean], index=["count","average"])
    df = pd.DataFrame(result)
    display(df)

productMarginLowerThanAverage()

0
count    59.000000
average   0.292203

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

In [33]:
def productsSoldButNeverReturned() : 
    order_products = pd.merge(order_details,product,left_on="PRODUCT_NUMBER",how="left",right_on="PRODUCT_NUMBER")
    quantity_filter = order_products["UNIT_PRICE"]>500
    order_products_quantity = order_products.loc[(quantity_filter),:]
    products_not_returned = order_products_quantity[~order_products_quantity["ORDER_DETAIL_CODE"].isin(returned_item["ORDER_DETAIL_CODE"])].drop_duplicates("PRODUCT_NUMBER")
    products_not_returned_filtered = products_not_returned.loc[:,["PRODUCT_NAME"]]
    print(products_not_returned_filtered.count())
    display(products_not_returned_filtered)


productsSoldButNeverReturned()

PRODUCT_NAME    13
dtype: int64


PRODUCT_NAME
33               Hailstorm Steel Irons
36            Hailstorm Titanium Irons
41          Lady Hailstorm Steel Irons
44       Lady Hailstorm Titanium Irons
47        Hailstorm Titanium Woods Set
51           Hailstorm Steel Woods Set
55   Lady Hailstorm Titanium Woods Set
58      Lady Hailstorm Steel Woods Set
132                          Star Dome
136                       Star Gazer 2
138                       Star Gazer 3
141                       Star Gazer 6
263                     Husky Rope 200

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

In [34]:
def employee_define_manager() : 
    for index,row in sales_staff.iterrows(): 
        isManager = sales_staff.at[index, "POSITION_EN"]

        if ("Manager" in isManager):
            sales_staff.at[index, "IsManager?"] = "Ja"
        else:
            sales_staff.at[index, "IsManager?"] = "Nee"

    sales_staff_filtered = sales_staff.loc[:,["LAST_NAME","IsManager?"]]
    sales_staff_filtered_ordered = sales_staff_filtered.sort_values(by=["LAST_NAME"])

    display(sales_staff_filtered_ordered)

employee_define_manager()


LAST_NAME IsManager?
32   Allessori        Nee
99      Barros        Nee
77     Belding         Ja
43   Bergström         Ja
49     Bretton         Ja
..         ...        ...
20   Wiesinger        Nee
57      Wilcox        Nee
9   Willoughby        Nee
29     Winkler        Nee
69      Wright         Ja

[102 rows x 2 columns]

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [35]:
from datetime import date
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [36]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 

*Vreemd verwoorde opdracht. Volgens mij moet kort <= 12 jaar, en lang >25 jaar zijn.*

In [45]:
def employee_yearsEmployed():
    for index,row in sales_staff.iterrows():
        years = sales_staff.at[index, "DATE_HIRED"]
        format = datetime.strptime(years, date_format)
        years_employed = format.year 
        employee = sales_staff.at[index,"DATE_HIRED"]

        if((date.today().year-years_employed)>=12):
            sales_staff.at[index,"DUUR_DIENST"] = "Lang in dienst"
        else :
            sales_staff.at[index,"DUUR_DIENST"] = "Kort in dienst"
   
    sales_staff_filtered = (sales_staff.loc[:,["LAST_NAME","DUUR_DIENST"]]).sort_values(by=["LAST_NAME"])
    display(sales_staff_filtered)

employee_yearsEmployed()
        

LAST_NAME     DUUR_DIENST
32   Allessori  Kort in dienst
99      Barros  Kort in dienst
77     Belding  Kort in dienst
43   Bergström  Kort in dienst
49     Bretton  Kort in dienst
..         ...             ...
20   Wiesinger  Kort in dienst
57      Wilcox  Kort in dienst
9   Willoughby  Kort in dienst
29     Winkler  Lang in dienst
69      Wright  Lang in dienst

[102 rows x 2 columns]